### DISCUSSION TITLE
Is this library usable with programming languages not C/C++?

### DISCUSSION THREAD

**Discussion by ghost** [frang75/nappgui_src#66] [General]
Because it relies on macros so much that I think it will not work without macros. There is the static/dynamic library available but it's not easy to use them with other languages not C/C++.

**Comment by ghost** [comment]
@frang75 This macro beast is too difficult to wrap that sometimes I think it's impossible to use outside of the C/C++ world at all. Good luck with #64 and #65.

**Comment by frang75** [comment]
As you may have seen, macros are for performing type checking at compile time. Any non-C/C++ language must call the function encapsulated in the macro. They usually end in _imp.

Functions in C are accessible through "almost" all programming languages.

**Reply by ghost** [reply]
Data structures specific to the library can only be created via these macros. So, yes. I'm able to call these functions from "almost" all programming languages, but how could I create the needed data structures from these languages, to be able to feed into these functions? Please let me know. Thank you.

**Comment by frang75** [comment]
`ArrSt(Product) *arrst_create(Product);`

`ArrSt(Product)` Is a macro that constructs a specific type for compile-time verification.
`arrst_create(Product)` Is a macro that calls array_create() with compile-time type checking.

Both use array.h which is the module that implements the API in C

```
_core_api Array *array_create(const uint16_t esize, const char_t *type);

_core_api Array *array_copy(const Array *array, FPtr_scopy func_copy, const char_t *type);
```

The rest of the types: Layout, Button, Slider, Panel, etc., are direct C types, where macros are not used.

**Reply by ghost** [reply]
Macros are used everywhere to create data structures. For example:

Check `DeclSt` in `core.hxx`. It is used to declare structures in `draw2d.hxx`, `geom2d.hxx` and `gui.hxx`. Check the definition of `DeclSt`, and it's using `ArrStDebug`, `SetStDebug`, `ArrStFuncs` and `SetStFuncs`.

Check `core/hfile.h`, these API are plain C, aren't them?

```
_core_api ArrSt(DirEntry) *hfile_dir_list(const char_t *pathname, const bool_t subdirs, ferror_t *error);

_core_api void hfile_dir_entry_remove(DirEntry *entry);
```

What about `DirEntry`? `DirEntry` is defined in `core.hxx` with:

```
typedef struct _direntry_t DirEntry;
DeclSt(DirEntry);
```

So, these plain C functions are expecting data structures that can only be created using macros.

**Reply by ghost** [reply]
@frang75

**Comment by frang75** [comment]
Dear @iahung2!

More than 10 years ago I started developing NAppGUI. I made the OpenSource project two years ago, so that anyone can use it freely. I continue to maintain the project in my free time, also for free. Therefore, you can do three things:
* Find another project that lives up to your expectations.
* Sponsor the project, as some companies do, hiring my services to urgently implement any feature you need.
* Download the NAppGUI book (also free), start programming and follow its numerous examples and detailed documentation. https://nappgui.com/ebook/ndoc_en.pdf

But:
If what you're looking for is to play Project Manager, or just get my attention for free, you're wrong. I advise you to look for a girlfriend, or boyfriend, or pet, or hobby. From now on you are banned from the project.

Thank you very much for your cooperation.
Sincerely.
